text/0000-task-builder.md

```diff
@@ -0,0 +1,200 @@
+
+Finally, a `schedule` method is provided to submit the task to the queue. For conveinance, it returns `JsUndefined`, matching Javascript semantics for most asynchronous methods.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Changes to low-level Task
+
+The existing low-level `Task` implementation in C++ is modified to match a new ownership model. The `perform` model will _consume_ self. The `complete` method must be changed to no longer pass `self` to prevent use-after-free.
+
+```rust
+pub fn Neon_Task_Schedule(task: *mut c_void,
+                          perform: unsafe extern fn(*mut c_void) -> *mut c_void,
+                          complete: unsafe extern fn(*mut c_void, &mut Local),
+                          callback: Local);
+```
+
+### Private `BaseTask` trait
+
+A new (private) `BaseTask` trait is introduced. The `BaseTask` trait is very similar to the existing `Task` trait except `fn perform` is changed to take ownership of `self` and `fn complete` is changed to an associated method (does not use self).
+
+A single `Output` parameter is provided and it is up to the user to return a `Result` if desired.
+
+```rust
+trait BaseTask: Send + Sized + 'static {
+    type Output: Send + 'static;
+    type JsEvent: Value;
+
+    fn perform(self) -> Self::Output;
+    fn complete<'a>(cx: TaskContext<'a>, result: Self::Output) -> JsResult<Self::JsEvent>;
+    fn schedule(self, callback: Handle<JsFunction>);
+}
+```
+
+### `Task` trait
+
+An implementation of `BaseTask` is added for `T: Task` to provide backwards compatibility to existing implementations of `Task`. Since `complete` is an associated method, `self` is threaded through the `Output` parameter.
+
+The existing `Task` trait will be _deprecated_.
+
+```rust
+impl<T: Task> BaseTask for T {
+    // Output contains a `Result` to match the `Task` trait
+    type Output = (Result<T::Output, T::Error>, T);
+    type JsEvent = T::JsEvent;
+
+    fn perform(self) -> Self::Output {
+        // `self` is threaded through `Output` as part of a tuple
+        (Task::perform(&self), self)
+    }
+
+    fn complete<'a>(cx: TaskContext<'a>, output: Self::Output) -> JsResult<Self::JsEvent> {
+        // `task` is unpacked from the tuple to use the `complete` method
+        let (result, task) = output;
+
+        task.complete(cx, result)
+    }
+}
+```
+
+### `TaskBuilder`
+
+The `TaskBuilder` struct contains two pieces of key data:
+
+* A mutable reference to `Context`
+    - Allows the conveinance of returning `JsUndefined` from `.schedule`
+    - The `N-API` implementation will require access to `Env`
+* The `task` closure to be performed on another thread
+
+#### Limitation
+
+Ideally we would have a more simple struct type and only call out the specific trait bounds on `fn schedule`.
```

Comment:
I think this section would be easier to understand if we had the real type declaration that this RFC is proposing documented above, to compare and contrast with.

Reply:
I'm going to try a little bit more to fix this issue before resigning to the more limited version.